### PR TITLE
Transition to weakdeps and requires for backend handling

### PR DESCRIPTION
**Pull request by luraess**
Change backend handling from deps to weakdeps to allow for conditional loading in Julia >= v1.9. For backward compatibility, use Requires.
The refactoring includes also
- examples for all supported backend
- API changes in Metal

EDITS:
- Add tests

### CODE REVIEW

File: src/CPUBackend.jl [hunk 1]
```diff
@@ -9,18 +9,18 @@ struct CPUEvent end
...
-    nblocks = ntuple(Val(N)) do I
+@inline function split(ndrange::CartesianIndices, nthreads::NTuple)
+    nblocks = ntuple(Val(length(nthreads))) do I
         return cld(size(ndrange, I), nthreads[I])
     end
     return nblocks
 end
 
-function (k::Kernel{<:CPUDevice})(args::Vararg{Any,B}; ndrange::NTuple{N,T}, priority=:low, nthreads=nothing) where {N,T,B}
+function (k::Kernel{<:CPUDevice})(args::Vararg{Any,B}; ndrange, priority=:low, nthreads=nothing) where {B}
```
Comment (luraess) [inline src/CPUBackend.jl]: Revert `::NTuple` annotation to allow for more general type of `ndrange`. 
@albert-de-montserrat feel free to check  if it's fine wrt type stability you were investigating.
@utkinis I was checking and it seems that prior to type stability fixes it did also fail if `ndrange=CartesianIndices(A)` in function call on `ndrange_to_indices()`

Comment (utkinis) [reply src/CPUBackend.jl]: @luraess Thanks for checking, added the missing method!

Comment (albert-de-montserrat) [reply src/CPUBackend.jl]: Looks good to me. It's able to do constant propagation on `ndims(::CartesianIndices)`
```julia-repl
julia> x
CartesianIndices((4:29, 4:29))

julia> @code_warntype ndims(x)
MethodInstance for ndims(::CartesianIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}})
  from ndims(R::CartesianIndices) @ Base.IteratorsMD multidimensional.jl:378
Arguments
  #self#::Core.Const(ndims)
  R::CartesianIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}}
Body::Int64
1 ─ %1 = Base.IteratorsMD.typeof(R)::Core.Const(CartesianIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}})
│   %2 = Base.IteratorsMD.ndims(%1)::Core.Const(2)
└──      return %2
```

**Comment by luraess** [comment]
@utkinis can you check whether the refactored examples are OK? If so, I could merge the PR and release the version 0.2.0

**Comment by luraess** [comment]
https://github.com/utkinis/TinyKernels.jl/pull/7/commits/3500037f86cf792157008d1e3b45a3f61a97b144 Looks good 👍

**Comment by utkinis** [comment]
Thanks @luraess for all the work on this. I think this PR is good for merging

**Pull request merged**